## mapDamage

[![Project Status: Inactive – The project has reached a stable, usable state but is no longer being actively developed; support/maintenance will be provided as time allows.](https://www.repostatus.org/badges/latest/inactive.svg)](https://www.repostatus.org/#inactive)

#### Regular installation

Refers to the detailed instructions on the [dedicated page](http://ginolhac.github.io/mapDamage/)

Briefly:

- Download a [release](https://github.com/ginolhac/mapDamage/releases) as `zip` or `tgz` file
- Decompress the archive
- Install both [R](http://www.r-project.org/) and the 4 mandatory packages.
- Install the python package `pysam`
- Install `mapDamage`, once in the archive folder, for a user: `python3 -m pip install --user .`

---

### Important

- From version `2.2.1` the `main` branch is requiring **python3** as `python2` is not supported from 2020-01-01.

- Users with versions dating prior to June 12 2013 please update. A nasty bug that caused the statistical part of `mapDamage` to use half of the data for estimation of the damage parameters, sorry for the inconvenience.

### Introduction

Complete documentation, instructions, examples, screenshots and FAQ are available at [this address](http://ginolhac.github.io/mapDamage/).

[mapDamage2](https://geogenetics.ku.dk/publications/mapdamage2.0/) is a computational framework written in **Python3** and **R**, which tracks and quantifies DNA damage patterns
among ancient DNA sequencing reads generated by Next-Generation Sequencing platforms.

`mapDamage` was developed at the [Centre for GeoGenetics](https://geogenetics.ku.dk/) by the [Orlando Group ](https://geogenetics.ku.dk/research_groups/palaeomix_group/).

### Citation

If you use this program, please cite the following publication:
Jónsson H, Ginolhac A, Schubert M, Johnson P, Orlando L.
[mapDamage2.0: fast approximate Bayesian estimates of ancient DNA damage parameters.
_Bioinformatics_ 23rd April 2013. doi: 10.1093/bioinformatics/btt193](http://bioinformatics.oxfordjournals.org/content/early/2013/05/17/bioinformatics.btt193)

The original `mapDamage1` was published in the following article:
Ginolhac A, Rasmussen M, Gilbert MT, Willerslev E, Orlando L.
[mapDamage: testing for damage patterns in ancient DNA sequences. _Bioinformatics_ 2011 **27**(15):2153-5
http://bioinformatics.oxfordjournals.org/content/27/15/2153](http://bioinformatics.oxfordjournals.org/content/27/15/2153)

### Contact

Please report bugs and suggest possible improvements on GitHub:
https://github.com/ginolhac/mapDamage/issues/new
